ktor-features/ktor-metrics-micrometer/jvm/src/io/ktor/metrics/micrometer/MicrometerMetrics.kt

```diff
@@ -0,0 +1,185 @@
+ *           <li><code>route</code>: The use ktor route used for this request. (e.g. '/some/path/{someParameter}')
+ *           <li><code>status</code>: The http status code or the Exception name thrown during the request (e.g. '200', 'java.lang.IllegalArgumentException')
+ *        <ul>
+ *     <li>
+ *  <ul>
+ */
+class MicrometerMetrics(
+    private val registry: MeterRegistry,
+    timerDistributionConfig: DistributionStatisticConfig,
+    private val tagCustomizer: MicrometerMetrics.TagBuilder.(ApplicationCall) -> Unit
+) {
+
+    private val active = registry.gauge("$baseName.requests.active", AtomicInteger(0))
+
+    init {
+        enableTimerDistributionConfig(timerDistributionConfig)
+    }
+
+    private fun enableTimerDistributionConfig(timerDistributionConfig: DistributionStatisticConfig) {
+        registry.config().meterFilter(object : MeterFilter {
+            override fun configure(id: Meter.Id, config: DistributionStatisticConfig): DistributionStatisticConfig =
+                if (id.name == requestTimerName)
+                    timerDistributionConfig.merge(config)
+                else
+                    config
+        })
+    }
+
+    /**
+     * Configures this Feature
+     * @property registry The meter registry where the meters are registered. Mandatory
+     * @property meterBinders The binders that are automatically bound to the registry. Default: [ClassLoaderMetrics],
+     * [JvmMemoryMetrics], [ProcessorMetrics], [JvmGcMetrics], [ProcessorMetrics], [JvmThreadMetrics], [FileDescriptorMetrics]
+     * @property distributionStatisticConfig configures the histogram and/or percentiles for all request timers. By
+     * default 50%, 90% , 95% and 99% percentiles are configured
+     * @property tagBuilder is called to customize the tags of the timer for each call. by default this is a no-op
+     */
+    class Configuration {
+
+        lateinit var registry: MeterRegistry
+
+        var meterBinders: List<MeterBinder> = listOf(
+            ClassLoaderMetrics(),
+            JvmMemoryMetrics(),
+            JvmGcMetrics(),
+            ProcessorMetrics(),
+            JvmThreadMetrics(),
+            FileDescriptorMetrics()
+        )
+
+        var distributionStatisticConfig: DistributionStatisticConfig =
+            DistributionStatisticConfig.Builder()
+                .percentiles(0.5, 0.9, 0.95, 0.99)
+                .build()
+
+        var tagCustomizer: TagBuilder.(ApplicationCall) -> Unit = {}
+    }
+
+    private fun ApplicationCall.getLocalAddress() = request.local.let { "${it.host}:${it.port}" }
+
+    private fun ApplicationCall.getStatusCode() = response.status()?.value?.toString() ?: "Unknown"
+
+    private fun ApplicationCall.tags(throwable: Throwable? = null) = listOf<Tag>(
+        Tag.of("local", getLocalAddress()),
+        Tag.of("method", request.httpMethod.value),
+        Tag.of("route", attributes[measureKey].route ?: request.path()),
```

Comment:
It's not clear should it be route spec or the actual request path? E.g. `/uri/{someParamter}` or `/uri/actualValue` ? Which one do we need or both?

Reply:
The actual one is not recommended (at least in prometheus), since it creates too many metrics (each different label value creates a separate metric to track). So when you have too many possible values in the url (like customer id of a tens of thousands of users...) it just bloats the metrics db and makes the searches slower.

Reply:
So if we have a handler out of the routing then we will always have actual path that seems to be reasonable fallback.